src/libraries/System.Net.Http/src/System/Net/Http/Headers/CacheControlHeaderParser.cs

```diff
@@ -22,7 +23,28 @@ private CacheControlHeaderParser()
         protected override int GetParsedValueLength(string value, int startIndex, object? storeValue,
             out object? parsedValue)
         {
-            CacheControlHeaderValue? temp = storeValue as CacheControlHeaderValue;
+            CacheControlHeaderValue? temp = null;
+            if (storeValue is not null)
+            {
+                if (storeValue is List<object> list)
+                {
+                    foreach (object item in list)
+                    {
+                        if (item is not HttpHeaders.InvalidValue)
+                        {
+                            temp = item as CacheControlHeaderValue;
+                            break;
+                        }
+                    }
+                }
+                else
+                {
+                    if (storeValue is not HttpHeaders.InvalidValue)
+                    {
+                        temp = storeValue as CacheControlHeaderValue;
+                    }
+                }
+            }
```

Comment:
```suggestion
            if (storeValue is List<object> list)
            {
                foreach (object item in list)
                {
                    if (item is CacheControlHeaderValue cacheControl)
                    {
                        temp = cacheControl;
                        break;
                    }
                }
            }
            else
            {
                temp = storeValue as CacheControlHeaderValue;
            }
```

Can skip the `InvalidValue` type checks here.

The below `Assert` should also be removed.